### PR TITLE
ParameterDict copy and __reversed__

### LINKED ISSUE #68476: ParameterDict should implement `.get` and `.setdefault`
## 🚀 Feature

`ParameterDict` implements most, but not all methods from `dict`. In particular it doesn't have:

    >>> set(dir({})) - set(dir(nn.ParameterDict()))
    {'__class_getitem__',
     '__ior__',
     '__or__',
     '__reversed__',
     '__ror__',
     'copy',
     'fromkeys',
     'get',
     'popitem',
     'setdefault'}

Arguably `copy`, `fromkeys` and `popitem` are not particularly useful, as well as the various `or` methods.

However `.get()` and `.setdefault()` can make code much nicer.

## Motivation

I'm implementing an optimizable optimizer with parameter specific moment and velocity.
The code looks something like this:

    def __init__(self):
        self.cache = nn.ParameterDict()
        ...
    def adjust(self, params):
        ...
        for name, param in params:
            m = self.cache.get('m', torch.zeros(param.shape))
            v = self.cache.get('v', torch.zeros(param.shape) + self.eps)
            ...

I can't define these parameter specific parameters in my `__init__`, since I only know the shape they need once the first optimization step has started.

## Pitch

The documentation writes "ParameterDict is an ordered dictionary".
It seems that as such it should implement the standard python dictionary methods.
In fact, it already implements nearly all of them, so why not those last two?

## Alternatives

I could write

    if 'm' not in self.cache:
            self.cache[name]['m'] = torch.zeros(param.shape)
    if 'v' not in self.cache 
            self.cache[name]['v'] = torch.zeros(param.shape) + self.eps
    m = self.cache[name]['m']
    v = self.cache[name]['v']

But it is 3 times as many lines of code, and any python developer would wonder "why doesn't this just use `.get` or `.default`?"

cc @albanD @mruberry @jbschlosser @walterddr

### PR DESCRIPTION
**Pull request by gramliu**
Partially resolves pytorch#68476 (comment)

* Added `copy` method
* Added `__reverse__` method

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -1834,6 +1834,26 @@ def check():
...
+
+        # Check copy works
+        copy = parameter_dict.copy()
+        parameter_dict["p15"] = Parameter(torch.randn(10, 10))
+        copy["p16"] = Parameter(torch.randn(9, 10))
+
+        self.assertEqual("p15" in parameter_dict, True)
+        self.assertEqual("p15" in copy, False)
+        self.assertEqual("p16" in parameter_dict, False)
+        self.assertEqual("p16" in copy, True)
```
Comment (jfrucht25) [inline test/test_nn.py]: I think you can just use assertTrue and assertFalse.
I would also go through and make sure all the other elements are the same between the old version and the copy.

**Review by jfrucht25 — CHANGES_REQUESTED** [review]
Looks good, just a couple small things.

File: test/test_nn.py (hunk 1 above)
Comment (gramliu) [reply test/test_nn.py]: Thanks for pointing this out. I went ahead and pushed a change to address this

**Review by jfrucht25 — APPROVED** [review]
looks good!

**Comment by shawnchan2014** [comment]
When I tried to integrate this PR into mines, the unit test actually fails at the `check()` at line 1862 because of `self.assertEqual(len(parameter_dict), len(parameters))` at line 1762. The reason is that `parameter_dict` is updated at line 1855 while `parameters` is not, causing the difference

**Comment by shawnchan2014** [comment]
Since this PR has already been approved, I added the two-liner fix in my PR <https://github.com/thisisalbertliang/pytorch/pull/7>

**Pull request merged**